include/seqan3/search/detail/search_configurator.hpp

```diff
@@ -100,26 +150,87 @@ class search_configurator
-                                                                policy_max_error,
-                                                                policy_search_result_builder<configuration_t>>;
-            return algorithm_t{cfg, index};
-        }
+        using search_result_t = typename select_search_result<configuration_t, index_t>::type;
+        using search_result_collection_t = std::vector<search_result_t>;
+        using type_erased_algorithm_t = std::function<search_result_collection_t(query_t)>;
+
+        return configure_hit_strategy<type_erased_algorithm_t>(cfg | search_cfg::detail::result_type<search_result_t>,
+                                                               index);
     }
+
+    template <typename algorithm_t, typename configuration_t, typename index_t>
+    static algorithm_t configure_hit_strategy(configuration_t const &, index_t const &);
 };
 
+/*!\brief Configures the algorithm with the correct hit strategy.
+ *
+ * \tparam algorithm_t The type erased algorithm used for the fixed return type.
+ * \tparam configuration_t The type of the search configuration.
+ * \tparam index_t The type of the index.
+ *
+ * \returns The configured search algorithm.
+ *
+ * \details
+ *
+ * If the algorithm was configured with the dynamic hit configuration element seqan3::search_cfg::hit, the
+ * configuration element is removed and replaced by the selected static hit configuration element.
+ * If the hit configuration element is already a static one nothing is changed in the configuration.
+ * After selecting the correct hit strategy the corresponding search algorithm is created with the new configuration
+ * and the given index.
+ *
+ * \throws std::invalid_argument if the dynamic hit configuration was not initialised with a hit strategy.
+ *
+ * If no hit configuration was configured a static assert is emitted during compilation.
+ */
+template <typename algorithm_t, typename configuration_t, typename index_t>
+algorithm_t search_configurator::configure_hit_strategy(configuration_t const & cfg, index_t const & index)
+{
+    // Select the actual search algorithm and return an instance constructed with the adapted config and the index.
+    auto generate_algorithm = [&] (auto new_cfg) -> algorithm_t
+    {
+        using new_configuration_t = decltype(new_cfg);
+        using selected_algorithm_t =
+            typename select_search_algorithm<new_configuration_t,
+                                             index_t,
+                                             policy_max_error,
+                                             policy_search_result_builder<new_configuration_t>>::type;
+
+        return selected_algorithm_t{new_cfg, index};
+    };
+
+    // Check if dynamic config present, otherwise continue.
+    if constexpr (configuration_t::template exists<search_cfg::hit>())
+    {
+        auto hit_variant = get<search_cfg::hit>(cfg).value;
+
+        if (std::holds_alternative<empty_type>(hit_variant))
```

Comment:
Hm I wonder if this is better than just setting letting the variant default to `hit_all` on initialisation as this is also the default when no hit configuration is given?

Reply:
For the configuartion we decided to not infer a default on the construction but let the algorithm handle the default. But even then I wasn't sure what is better in this situation. So currently we infer some default if the user provides nothing. But the default would be the slowest configuration finding all hits. As far as I know this is also something that you don't want in many scenarios. The question here is the same with the alignment. Whether global or local should be default is very dependent on the use case and nothing we should default but require from the user to be set. Now the user explicitly wants to set the dynamic hit strategy but made a mistake in the code and actually never sets anything but doesn't know it (for example he would like to never use all_hits strategy). In this case we would assume a default which might be something the user really don't want and he/she either get's wrong results or is wondering about the performance or memory consumption etc. So from that perspective I think it would be wiser to not assume a default configuration here but let the user know that the search was misconfigured. Defaults are only good if they do the right thing for the user, but I don't see this here.